client/servers/manager.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// FindServer returns a server to send an RPC too. If there are no servers, nil
+// is returned.
+func (m *Manager) FindServer() *Server {
+	m.Lock()
+	defer m.Unlock()
+
+	if len(m.servers) == 0 {
+		m.logger.Printf("[WARN] manager: No servers available")
+		return nil
+	}
+
+	// Return whatever is at the front of the list because it is
+	// assumed to be the oldest in the server list (unless -
+	// hypothetically - the server list was rotated right after a
+	// server was added).
+	return m.servers[0]
+}
+
+// NumNodes returns the number of approximate nodes in the cluster.
+func (m *Manager) NumNodes() int32 {
+	m.Lock()
+	defer m.Unlock()
+	return m.numNodes
+}
+
+// SetNumNodes stores the number of approximate nodes in the cluster.
+func (m *Manager) SetNumNodes(n int32) {
+	m.Lock()
+	defer m.Unlock()
+	m.numNodes = n
+}
+
+// NotifyFailedServer marks the passed in server as "failed" by rotating it
+// to the end of the server list.
+func (m *Manager) NotifyFailedServer(s *Server) {
+	m.Lock()
+	defer m.Unlock()
+
+	// If the server being failed is not the first server on the list,
+	// this is a noop.  If, however, the server is failed and first on
+	// the list, move the server to the end of the list.
+	if len(m.servers) > 1 && m.servers[0] == s {
+		m.servers.cycle()
+	}
+}
+
+// NumServers returns the total number of known servers whether healthy or not.
+func (m *Manager) NumServers() int {
+	m.Lock()
+	defer m.Unlock()
+	return len(m.servers)
+}
+
+// GetServers returns a copy of the current list of servers.
+func (m *Manager) GetServers() Servers {
+	m.Lock()
+	defer m.Unlock()
+
+	copy := make([]*Server, 0, len(m.servers))
+	for _, s := range m.servers {
+		ns := new(Server)
+		*ns = *s
+		copy = append(copy, ns)
+	}
+
+	return copy
+}
+
+// RebalanceServers shuffles the order in which Servers will be contacted. The
+// function will shuffle the set of potential servers to contact and then attempt
+// to contact each server. If a server successfully responds it is used, otherwise
+// it is rotated such that it will be the last attempted server.
+func (m *Manager) RebalanceServers() {
+	m.Lock()
+	defer m.Unlock()
+
+	// Shuffle servers so we have a chance of picking a new one.
+	m.servers.shuffle()
+
+	// Iterate through the shuffled server list to find an assumed
+	// healthy server.  NOTE: Do not iterate on the list directly because
+	// this loop mutates the server list in-place.
+	var foundHealthyServer bool
+	for i := 0; i < len(m.servers); i++ {
+		// Always test the first server.  Failed servers are cycled
+		// while Serf detects the node has failed.
+		srv := m.servers[0]
+
+		err := m.connPoolPinger.Ping(srv.Addr)
+		if err == nil {
+			foundHealthyServer = true
+			break
+		}
+		m.logger.Printf(`[DEBUG] manager: pinging server "%s" failed: %s`, srv, err)
+
+		m.servers.cycle()
+	}
+
+	if !foundHealthyServer {
+		m.logger.Printf("[DEBUG] manager: No healthy servers during rebalance, aborting")
+		return
```

Comment:
Don't need this `return` and the word `aborting` in the log line is a bit misleading. We're not actually aborting anything here, we just couldn't find a healthy server while rebalancing.